pkg/engine/task/pod_exec.go

```diff
@@ -0,0 +1,232 @@
+// ContainerName - optional container to execute the command in. If empty, first container is taken
+// Args - The command (and args) to execute.
+// In - Command input stream.
+// Out - Command output stream
+// Err - Command error stream
+type PodExec struct {
+	RestCfg       *rest.Config
+	PodName       string
+	PodNamespace  string
+	ContainerName string
+	Args          []string
+	In            io.Reader
+	Out           io.Writer
+	Err           io.Writer
+	TTY           bool
+}
+
+// Run executes a command in a pod. This is a distilled version of what `kubectl exec` (and
+// also `kubectl  cp`) doing under the hood: a POST request is made to the `exec` subresource
+// of the v1/pods endpoint containing the pod information and the command. Here is a good article
+// describing it in detail: https://erkanerol.github.io/post/how-kubectl-exec-works/
+func (pe *PodExec) Run() error {
+	codec := serializer.NewCodecFactory(scheme.Scheme)
+	restClient, err := apiutil.RESTClientForGVK(
+		schema.GroupVersionKind{
+			Version: "v1",
+			Kind:    "pods",
+		},
+		pe.RestCfg,
+		codec)
+	if err != nil {
+		return err
+	}
+
+	req := restClient.
+		Post().
+		Resource("pods").
+		Name(pe.PodName).
+		Namespace(pe.PodNamespace).
+		SubResource("exec")
+
+	req.VersionedParams(&v1.PodExecOptions{
+		Stdin:     pe.In != nil,
+		Stdout:    pe.Out != nil,
+		Stderr:    pe.Err != nil,
+		TTY:       pe.TTY,
+		Container: pe.ContainerName,
+		Command:   pe.Args,
+	}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(pe.RestCfg, "POST", req.URL())
+	if err != nil {
+		return err
+	}
+
+	so := remotecommand.StreamOptions{
+		Stdin:  pe.In,
+		Stdout: pe.Out,
+		Stderr: pe.Err,
+		Tty:    pe.TTY,
+	}
+
+	// The result of the executor.Stream() call itself is returned through the streams (In, Out and Err)
+	// defined in the PodExec, e.g. when downloading a file, pe.Out will  contain the file bytes.
+	return exec.Stream(so)
+}
+
+// FileSize fetches the size of a file in a remote pod. It runs `stat -c %s file` command in the
+// pod and parses the output.
+func FileSize(file string, pod *v1.Pod, restCfg *rest.Config) (int64, error) {
+	readout, stdout := io.Pipe()
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"stat", "-c", "%s", file},
+		In:            nil,
+		Out:           stdout,
+		Err:           nil,
+		TTY:           true, // this will forward 2>&1. otherwise, reading from Out will never return for e.g. missing files
+	}
+
+	errCh := make(chan error, 1)
+	go func() {
+		defer stdout.Close() // Close never returns an error
+		errCh <- pe.Run()
+	}()
+
+	buf, err := ioutil.ReadAll(readout)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get the size of %s `: %v", file, err)
+	}
+
+	raw := string(buf)
+
+	// THIS CHECK HAS TO HAPPEN AFTER WE CONSUME THE STDOUT/STDERR READER.
+	if err := <-errCh; err != nil {
```

Comment:
Why must `pe.Run()` run in a Goroutine? Can't you just wait for it to finish, and use a [strings.Builder](https://golang.org/pkg/strings/#Builder) on `pe.Out` to get the task output?

Reply:
I guess I was so blinded by the `io.Pipe` synchronous nature that I went too deep down this rabbit hole. You're right - we don't need it here, and then we don't need the goroutine. `strings.Builder` works just fine 🎉